services/users.js

```diff
@@ -1,44 +1,22 @@
-const { USERS_PATCH_HANDLER_SUCCESS_MESSAGES, USERS_PATCH_HANDLER_ERROR_MESSAGES } = require("../constants/users");
 const firestore = require("../utils/firestore");
 const { formatUsername } = require("../utils/username");
 const userModel = firestore.collection("users");
-const archiveUsers = async (usersData) => {
-  const batch = firestore.batch();
-  const usersBatch = [];
-  const summary = {
-    totalUsersArchived: 0,
-    totalOperationsFailed: 0,
-    updatedUserDetails: [],
-    failedUserDetails: [],
-  };
-
-  usersData.forEach((user) => {
-    const { id, first_name: firstName, last_name: lastName } = user;
-    const updatedUserData = {
-      ...user,
-      roles: {
-        ...user.roles,
-        archived: true,
-      },
-      updated_at: Date.now(),
-    };
-    batch.update(userModel.doc(id), updatedUserData);
-    usersBatch.push({ id, firstName, lastName });
-  });
+const tasksQuery = require("../models/tasks");
 
+const getUsersWithIncompleteTasks = async (users) => {
+  if (users.length === 0) return [];
   try {
-    await batch.commit();
-    summary.totalUsersArchived += usersData.length;
-    summary.updatedUserDetails = [...usersBatch];
-    return {
-      message: USERS_PATCH_HANDLER_SUCCESS_MESSAGES.ARCHIVE_USERS.SUCCESSFULLY_COMPLETED_BATCH_UPDATES,
-      ...summary,
-    };
-  } catch (err) {
-    logger.error("Firebase batch Operation Failed!");
-    summary.totalOperationsFailed += usersData.length;
-    summary.failedUserDetails = [...usersBatch];
-    return { message: USERS_PATCH_HANDLER_ERROR_MESSAGES.ARCHIVE_USERS.BATCH_DATA_UPDATED_FAILED, ...summary };
+    const eligibleUsersWithTasks = [];
+    for (const user of users) {
+      const abandonedTasksQuerySnapshot = await tasksQuery.fetchIncompleteTaskForUser(user.id);
+      if (!abandonedTasksQuerySnapshot.empty) {
+        eligibleUsersWithTasks.push(user);
+      }
+    }
+    return eligibleUsersWithTasks;
+  } catch (error) {
+    logger.error(`Error in getting users who abandoned tasks:  ${error}`);
+    throw error;
```

Comment:
We should use batch queries to firestore wherever possible. For example, fetch all incomplete tasks and then filter them by user IDs in memory.

Reply:
Okay, will update the implementation as suggested.